### ISSUE TITLE
Failure to type check object literals

### ISSUE THREAD

**Issue by phistuck** [google/closure-compiler#1502]
Looks like object literals are not type checked. :(
The following code does not generate any warning -
```
// ==ClosureCompiler==
// @output_file_name default.js
// @compilation_level ADVANCED_OPTIMIZATIONS
// @language ECMASCRIPT6_STRICT
// ==/ClosureCompiler==

var /** @type {Object<string, number>} */ Patterns =
 {
  PATTERN1: /expression.*/,
  PATTERN2: "hey"
 };

var /** @type {Object<string, RegExp>} */ Patterns1 =
 {
  PATTERN1: /expression.*/,
  PATTERN2: "hey"
 };

var /** @const {Object<string, number>} */ Patterns2 =
 {
  PATTERN1: /expression.*/,
  PATTERN2: "hey"
 };

var /** @const {Object<string, RegExp>} */ Patterns3 =
 {
  PATTERN1: /expression.*/,
  PATTERN2: "hey"
 };

const /** @type {Object<string, number>} */ Patterns4 =
 {
  PATTERN1: /expression.*/,
  PATTERN2: "hey"
 };

const /** @type {Object<string, RegExp>} */ Patterns5 =
 {
  PATTERN1: /expression.*/,
  PATTERN2: "hey"
 };

const /** @const {Object<string, number>} */ Patterns6 =
 {
  PATTERN1: /expression.*/,
  PATTERN2: "hey"
 };

const /** @const {Object<string, RegExp>} */ Patterns7 =
 {
  PATTERN1: /expression.*/,
  PATTERN2: "hey"
 };


console.log(Patterns, Patterns1, Patterns2, Patterns3, Patterns4, Patterns5, Patterns6, Patterns7);
```

http://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540language%2520ECMASCRIPT6_STRICT%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250Avar%2520%252F**%2520%2540type%2520%257BObject%253Cstring%252C%2520number%253E%257D%2520*%252F%2520Patterns%2520%253D%250A%2520%257B%250A%2520%2520PATTERN1%253A%2520%252Fexpression.*%252F%252C%250A%2520%2520PATTERN2%253A%2520%2522hey%2522%250A%2520%257D%253B%250A%250Avar%2520%252F**%2520%2540type%2520%257BObject%253Cstring%252C%2520RegExp%253E%257D%2520*%252F%2520Patterns1%2520%253D%250A%2520%257B%250A%2520%2520PATTERN1%253A%2520%252Fexpression.*%252F%252C%250A%2520%2520PATTERN2%253A%2520%2522hey%2522%250A%2520%257D%253B%250A%250Avar%2520%252F**%2520%2540const%2520%257BObject%253Cstring%252C%2520number%253E%257D%2520*%252F%2520Patterns2%2520%253D%250A%2520%257B%250A%2520%2520PATTERN1%253A%2520%252Fexpression.*%252F%252C%250A%2520%2520PATTERN2%253A%2520%2522hey%2522%250A%2520%257D%253B%250A%250Avar%2520%252F**%2520%2540const%2520%257BObject%253Cstring%252C%2520RegExp%253E%257D%2520*%252F%2520Patterns3%2520%253D%250A%2520%257B%250A%2520%2520PATTERN1%253A%2520%252Fexpression.*%252F%252C%250A%2520%2520PATTERN2%253A%2520%2522hey%2522%250A%2520%257D%253B%250A%250Aconst%2520%252F**%2520%2540type%2520%257BObject%253Cstring%252C%2520number%253E%257D%2520*%252F%2520Patterns4%2520%253D%250A%2520%257B%250A%2520%2520PATTERN1%253A%2520%252Fexpression.*%252F%252C%250A%2520%2520PATTERN2%253A%2520%2522hey%2522%250A%2520%257D%253B%250A%250Aconst%2520%252F**%2520%2540type%2520%257BObject%253Cstring%252C%2520RegExp%253E%257D%2520*%252F%2520Patterns5%2520%253D%250A%2520%257B%250A%2520%2520PATTERN1%253A%2520%252Fexpression.*%252F%252C%250A%2520%2520PATTERN2%253A%2520%2522hey%2522%250A%2520%257D%253B%250A%250Aconst%2520%252F**%2520%2540const%2520%257BObject%253Cstring%252C%2520number%253E%257D%2520*%252F%2520Patterns6%2520%253D%250A%2520%257B%250A%2520%2520PATTERN1%253A%2520%252Fexpression.*%252F%252C%250A%2520%2520PATTERN2%253A%2520%2522hey%2522%250A%2520%257D%253B%250A%250Aconst%2520%252F**%2520%2540const%2520%257BObject%253Cstring%252C%2520RegExp%253E%257D%2520*%252F%2520Patterns7%2520%253D%250A%2520%257B%250A%2520%2520PATTERN1%253A%2520%252Fexpression.*%252F%252C%250A%2520%2520PATTERN2%253A%2520%2522hey%2522%250A%2520%257D%253B%250A%250A%250Aconsole.log(Patterns%252C%2520Patterns1%252C%2520Patterns2%252C%2520Patterns3%252C%2520Patterns4%252C%2520Patterns5%252C%2520Patterns6%252C%2520Patterns7)%253B

**Comment by marcfallows** [comment]
I was just about to report the same bug:

https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%250A%2520*%2520%2540param%2520%257B!Object%253Cstring%252C%2520boolean%253E%257D%2520o%250A%2520*%252F%250Afunction%2520f(o)%2520%257B%250A%2520%2520console.log(o)%253B%250A%257D%250A%250Af(%257B'foo'%253A%2520true%257D)%253B%250Af(%257B'foo'%253A%2520'bar'%257D)%253B%250Af(%257Bfoo%253A%2520true%257D)%253B%250Af(%257Bfoo%253A%2520'bar'%257D)%253B%250A%250A%252F**%2520%2540type%2520%257B!Object%253Cstring%253E%257D%2520*%252F%250Aconst%2520o%2520%253D%2520%257B%257D%253B%250Ao%255B'foo'%255D%2520%253D%2520true%253B%250Ao%255B'foo'%255D%2520%253D%2520'bar'%253B

This has been sitting around for a while though...
